### PR TITLE
Add Network reusable component

### LINKED ISSUE #393: VCH reconfigure UI - Network
**User Statement:**
As a VIC User, I would like to configure an already created VCH. Particularly step "Network"

The mockups are continuously updated at this moment but can be accessed here.

**Acceptance Criteria:**

User should be able to access the view info and the edit modal of a VCH by clicking on the left side navigation of the configure view tab of a VCH. Particularly step "Network"

**Details:**

Implementation the reconfiguration of VCH based on:

https://github.com/vmware/vic-ux/issues/76

and 

https://github.com/vmware/vic/issues/6123

Verify #279 and #280 have been implemented as a requirement for this feature

### PR DESCRIPTION
**Pull request by javierfz1980**
This commit will introduce several changes to the VCH Network component in order to make it reusable on Configure View (read only) / VCH Configure Modal (edit mode) and VCH Creation Wizard.

Fixes #393 

PR acceptance checklist:

[ x ] All unit tests pass
[ x ] All e2e tests pass
[ x ] Unit test(s) included*
[ n/a ] e2e test(s) included*
[ x ] Screenshot attached and UX approved*

Network vch create wizard
network-create

Network reconfigure
network-reconfigure

Network reconfigure requests
network-reconfigure-request

### CODE REVIEW

File: h5c/vic/src/vic-webapp/src/app/shared/utils/vch/vch-utils.ts [hunk 1]
```diff
@@ -155,25 +156,18 @@ export function processPayloadFromUiToApi(payload): VchApi {
...
-
-        if (payload.networks.dnsServer && payload.networks.dnsServer.length) {
-          vchNetwork.client.nameservers = payload.networks.dnsServer;
+          vchNetwork.client.gateway.routing_destinations = payload.networks.clientNetworkRouting.split(',');
         }
       }
     }
 
-    if (payload.networks.managementNetwork) {
+    if (payload.networks.managementNetwork && payload.networks.managementNetwork !== '') {
```
Comment (jak-atx) [inline h5c/vic/src/vic-webapp/src/app/shared/utils/vch/vch-utils.ts]: wouldn't checking if(payload.networks.managementNetwork) suffice for checking empty string?

File: h5c/vic/src/vic-webapp/src/app/shared/utils/vch/vch-utils.ts [hunk 2]
```diff
@@ -414,12 +411,51 @@ export function processPayloadFromApiToUi (vch: VchApi): VchUi {
           volFileFolder: getDSNameAndFolder(vol.datastore).folderPath
         })
       }),
-
+    },
+    networks: {
+      bridgeNetwork: vch.network.bridge.port_group.id,
+      bridgeNetworkRange: vch.network.bridge.ip_range, // '172.16.0.0/12',
```
Comment (jak-atx) [inline h5c/vic/src/vic-webapp/src/app/shared/utils/vch/vch-utils.ts]: is there any reason for the comment here?

File: h5c/vic/src/vic-webapp/src/app/shared/utils/vch/vch-utils.ts [hunk 3]
```diff
@@ -414,12 +411,51 @@ export function processPayloadFromApiToUi (vch: VchApi): VchUi {
...
+      managementNetwork: vch.network.management.port_group.id,
+      managementNetworkIp: vch.network.management.gateway ? vch.network.management.gateway.routing_destinations.join(',') : '',
+      managementNetworkType: vch.network.management.nameservers && vch.network.management.nameservers.length > 0 ? 'static' : 'dhcp',
+      managementNetworkGateway: vch.network.management.gateway ? vch.network.management.gateway.address : '',
+      managementNetworkRouting: vch.network.management.nameservers.join(','),
+      containerNetworks: vchNetworkContainerApiToUi(vch.network.container),
+      httpProxy: vch.network.httpProxy ? vch.network.httpProxy.split(':')[0] : '',
+      httpProxyPort: vch.network.httpProxy ? vch.network.httpProxy.split(':')[1] : '',
+      httpsProxy: vch.network.httpsProxy ? vch.network.httpsProxy.split(':')[0] : '',
+      httpsProxyPort: vch.network.httpsProxy ? vch.network.httpsProxy.split(':')[1] : '',
```
Comment (jak-atx) [inline h5c/vic/src/vic-webapp/src/app/shared/utils/vch/vch-utils.ts]: this seems confusing because the payload and the interface seems to specify 'networks' as the object name rather than 'network'..so why is this vch.network rather than vch.networks?

File: h5c/vic/src/vic-webapp/src/app/interfaces/vch.ts [hunk 4]
```diff
@@ -131,36 +142,71 @@ export interface VchApiNetwork {
...
-  volume_stores?: [
-    {
-      datastore: string,
-      label: string
-    }
-    ],
+  volume_stores?: {
+    datastore: string;
+    label: string
+    }[],
```
Comment (jak-atx) [inline h5c/vic/src/vic-webapp/src/app/interfaces/vch.ts]: any reason to change this from the format it was before?

File: h5c/vic/src/vic-webapp/src/app/shared/components/vch-network/vch-network.component.html [hunk 5]
```diff
@@ -0,0 +1,823 @@
...
+            {{ model.managementNetwork }}, {{ model.managementNetworkIp }}
+          </ng-template>
+        </clr-stack-content>
+        <clr-stack-block>
+          <clr-stack-label>{{ i18n.translate('vch.storage.network.portgroupName') }}</clr-stack-label>
+          <clr-stack-content>{{ model.managementNetwork }}</clr-stack-content>
+        </clr-stack-block>
+        <clr-stack-block *ngIf="model.managementNetworkType === 'dhcp'; else managementStaticInfo">
+          <clr-stack-label>DHCP</clr-stack-label>
+          <clr-stack-content></clr-stack-content>
```
Comment (jak-atx) [inline h5c/vic/src/vic-webapp/src/app/shared/components/vch-network/vch-network.component.html]: Does this portion get dynamically populated?

File: h5c/vic/src/vic-webapp/src/app/shared/components/vch-network/vch-network.component.html [hunk 6]
```diff
@@ -0,0 +1,823 @@
...
+          </div>
+          <div class="col-xs text-right">
+            <label for="bridge-network-range-selector">{{ i18n.translate('vch.storage.network.bridgeNetworkRange') }}</label>
+          </div>
+          <div class="col-xs">
+            <label for="bridge-network-range-selector"
+                   aria-haspopup="true"
+                   role="tooltip"
+                   class="tooltip tooltip-validation tooltip-lg tooltip-top-left"
+                   [class.invalid]="form.get('bridgeNetworkRange').invalid && (form.get('bridgeNetworkRange').dirty || form.get('bridgeNetworkRange').touched)">
```
Comment (jak-atx) [inline h5c/vic/src/vic-webapp/src/app/shared/components/vch-network/vch-network.component.html]: I thought you had written a util function to check for isInvalid? If so, it seems could be used here and other places in this template instead of inline.

File: h5c/vic/src/vic-webapp/src/app/shared/components/vch-network/vch-network.component.html [hunk 7]
```diff
@@ -0,0 +1,823 @@
...
+          <div class="col-xs"></div>
+        </div>
+      </section>
+
+      <!-- Optional Container Networks -->
+      <section class="form-block" *ngIf="inAdvancedMode">
+        <label>{{ i18n.translate('vch.storage.network.optionalContainerNetworks') }}</label>
+        <div class="mb-2"
+             formArrayName="containerNetworks"
+             *ngFor="let containerNetwork of form.get('containerNetworks').controls; let i=index">
```
Comment (jak-atx) [inline h5c/vic/src/vic-webapp/src/app/shared/components/vch-network/vch-network.component.html]: why not just reuse 'index', instead of re-assigning to 'i'?

File: h5c/vic/src/vic-webapp/src/styles.scss [hunk 8]
```diff
@@ -15,3 +15,9 @@
 .cursor-pointer {
   cursor: pointer;
 }
+
+/* Fixes issue with clr-stack-block when some row bottom borders are not visible */
+.stack-view .stack-block .stack-block,
+.stack-view .stack-block {
```
Comment (jak-atx) [inline h5c/vic/src/vic-webapp/src/styles.scss]: are both declarations needed? it seems like just '.stack-view .stack-block' would cover it.

File: h5c/vic/src/vic-webapp/src/app/create-vch-wizard/networks/networks.component.ts [hunk 9]
```diff
@@ -1,345 +0,0 @@
-/*
```
[the remark this answers is not in the document]
Comment (javierfz1980) [reply h5c/vic/src/vic-webapp/src/app/create-vch-wizard/networks/networks.component.ts]: this is the new network component:

https://github.com/javierfz1980/vic-ui/blob/b532beb9c24d99d004c25b2da5783493568584f7/h5c/vic/src/vic-webapp/src/app/shared/components/vch-network/vch-network.component.ts

during refactor almost all the creation wizard core component has been removed from here:
https://github.com/vmware/vic-ui/tree/feature/vch-configure/vch-configure/h5c/vic/src/vic-webapp/src/app/create-vch-wizard

to here:
https://github.com/vmware/vic-ui/tree/feature/vch-configure/vch-configure/h5c/vic/src/vic-webapp/src/app/shared/components

the same happened on previous PRs on this feature branch

File: h5c/vic/src/vic-webapp/src/app/shared/utils/vch/vch-utils.ts (hunk 3 above)
Comment (javierfz1980) [reply h5c/vic/src/vic-webapp/src/app/shared/utils/vch/vch-utils.ts]: This is because they are two different interfaces:

"VchApi" is the interface for the vch payload retrieved from the API requests. I've made this interface based on the api documentation http://prntscr.com/jmb2jh

export interface VchApi {
  version?: string,
  name?: string,
  debug?: number,
  syslog_addr?: string
  compute?: VchApiCompute,
  network?: VchApiNetwork,
  storage?: VchApiStorage,
  auth?: VchApiAuth,
  endpoint?: VchApiEndpoint,
  registry?: VchApiRegistry,
  container?: VchApiContainer
  runtime?: VchApiRuntime,
}

"VchUi" is the interface to handle the vch information inside our UI. I've made this interface based on each of the creation wizard components that we already have on master, and also based on how they handle the vch information on their internal logic

https://github.com/vmware/vic-ui/blob/e8e3267cacb84b3e60770c43dbe0dc0dfc04341c/h5c/vic/src/vic-webapp/src/app/create-vch-wizard/storage-capacity/storage-capacity.component.ts#L177
https://github.com/vmware/vic-ui/blob/e8e3267cacb84b3e60770c43dbe0dc0dfc04341c/h5c/vic/src/vic-webapp/src/app/create-vch-wizard/storage-capacity/storage-capacity.component.ts#L177
https://github.com/vmware/vic-ui/blob/e8e3267cacb84b3e60770c43dbe0dc0dfc04341c/h5c/vic/src/vic-webapp/src/app/create-vch-wizard/networks/networks.component.ts#L339

export interface VchUi {
  general?: VchUiGeneral;
  computeCapacity?: VchUiCompute;
  storageCapacity?: VchUiStorage;
  networks?: VchUiNetwork;
  operations?: any, // work in progress
  registry?: any, // work in progress
  security?: any // work in progress
}

So basically we have two interfaces, and we have to convert the payloads between them:
1- VchApi -> is the api model and it has to be converted to VchUi after each vch GET request response.
2- VchUi -> is a more convenient structure to handle the information through our creation / configure flows and it has to be converted to VchApi before each vch POST / PUT / PATCH request.

In this particular case, we are converting the VchApi model -> "network" that comes from the API (http://prntscr.com/jmb2jh) into our internal VchUi model -> "networks" payload. The same happen for VchApi.compute -> VchUi.computeCapacity , VchApi.storage -> VchUi.storageCapacity , etc...

File: h5c/vic/src/vic-webapp/src/app/interfaces/vch.ts (hunk 4 above)
Comment (javierfz1980) [reply h5c/vic/src/vic-webapp/src/app/interfaces/vch.ts]: Because the format it was before was wrong... It was introduced by me, but I didn't used that interface till this network commit when I figured out that it was wrong. 

I wanted to make it an array of this kind of objects -> {datastore: string; label: string}

So the correct way to do it is {datastore: string; label: string}[] and not [{datastore: string; label: string}]

File: h5c/vic/src/vic-webapp/src/app/shared/components/vch-network/vch-network.component.html (hunk 5 above)
Comment (javierfz1980) [reply h5c/vic/src/vic-webapp/src/app/shared/components/vch-network/vch-network.component.html]: not sure what you mean... if you mean the 'clr-stack-content' portion... it should be empty because the DHCP doesn't have any ip value....  as opposed if it is an Static IP the 'clr-stack-content' is populated. You can check it a couple of lines below.

File: h5c/vic/src/vic-webapp/src/app/shared/components/vch-network/vch-network.component.html (hunk 7 above)
Comment (javierfz1980) [reply h5c/vic/src/vic-webapp/src/app/shared/components/vch-network/vch-network.component.html]: Most of the code of the template was copied from the original one, so I didn't change it:
https://github.com/vmware/vic-ui/blob/e8e3267cacb84b3e60770c43dbe0dc0dfc04341c/h5c/vic/src/vic-webapp/src/app/create-vch-wizard/networks/networks.html#L611

as a side note, AFAIK you have to declare the variable to get access to the index on an *ngFor . There is no way to get access to the index if you don't assign it to a variable.

File: h5c/vic/src/vic-webapp/src/app/shared/utils/vch/vch-utils.ts (hunk 3 above)
Comment (jak-atx) [reply h5c/vic/src/vic-webapp/src/app/shared/utils/vch/vch-utils.ts]: Ok, it seems one is model and one is a viewModel ..we should think about the naming of these interfaces to be more intuitive or add some comment explaining what each one is for.

File: h5c/vic/src/vic-webapp/src/app/shared/components/cli-command/cli-command.component.ts [hunk 10]
```diff
@@ -207,8 +208,9 @@ export class CliCommandComponent implements OnInit {
         results['storageCapacity']['imageStore'] + (results['storageCapacity']['fileFolder'] || '');
       delete results['storageCapacity']['fileFolder'];
 
-      results['storageCapacity']['baseImageSize'] = results['storageCapacity']['baseImageSize']
-        + results['storageCapacity']['baseImageSizeUnit'].replace('i', '');
+      const sizeUnit = results['storageCapacity']['baseImageSizeUnit'] ?
+        results['storageCapacity']['baseImageSizeUnit'].replace('i', '') : '';
```
[the remark this answers is not in the document]
Comment (jak-atx) [reply h5c/vic/src/vic-webapp/src/app/shared/components/cli-command/cli-command.component.ts]: Ok, we should think about putting that value in a common constant if we do need to change it.
